src/main/java/blusunrize/immersiveengineering/common/blocks/metal/ChuteTileEntity.java

```diff
@@ -0,0 +1,341 @@
+import net.minecraft.entity.item.ItemEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.state.EnumProperty;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.tileentity.TileEntityType;
+import net.minecraft.util.Direction;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.World;
+import net.minecraftforge.common.model.TRSRTransformation;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.items.IItemHandler;
+import net.minecraftforge.items.ItemHandlerHelper;
+
+import javax.vecmath.Matrix4f;
+import java.util.*;
+
+public class ChuteTileEntity extends IEBaseTileEntity implements IStateBasedDirectional, IAdvancedHasObjProperty,
+		IHammerInteraction, IAdvancedSelectionBounds, IAdvancedCollisionBounds
+{
+	private static final String NBT_POS = "immersiveengineering:chutePos";
+	private static final String NBT_TIME = "immersiveengineering:chuteTime";
+	private static final String NBT_GLITCH = "immersiveengineering:chuteGlitched";
+	public static TileEntityType<ChuteTileEntity> TYPE;
+
+	private boolean diagonal = false;
+
+	public ChuteTileEntity()
+	{
+		super(TYPE);
+	}
+
+	@Override
+	public EnumProperty<Direction> getFacingProperty()
+	{
+		return IEProperties.FACING_HORIZONTAL;
+	}
+
+	@Override
+	public PlacementLimitation getFacingLimitation()
+	{
+		return PlacementLimitation.HORIZONTAL;
+	}
+
+	@Override
+	public boolean mirrorFacingOnPlacement(LivingEntity placer)
+	{
+		return false;
+	}
+
+	@Override
+	public boolean canHammerRotate(Direction side, Vec3d hit, LivingEntity entity)
+	{
+		return !entity.isSneaking();
+	}
+
+	@Override
+	public boolean canRotate(Direction axis)
+	{
+		return true;
+	}
+
+	@Override
+	public void onEntityCollision(World world, Entity entity)
+	{
+		boolean contact = false;
+		Direction facing = getFacing();
+		if(!diagonal)
+			contact = entity.posY-getPos().getY() <= .125;
+		else if(facing==Direction.NORTH)
+			contact = entity.posZ-getPos().getZ() <= .125;
+		else if(facing==Direction.SOUTH)
+			contact = entity.posZ-getPos().getZ() >= .875;
+		else if(facing==Direction.WEST)
+			contact = entity.posX-getPos().getX() <= .125;
+		else if(facing==Direction.EAST)
+			contact = entity.posX-getPos().getX() >= .875;
+
+		if(this.diagonal&&entity.posY-getPos().getY() <= .625)
+		{
+			BlockPos target = getPos().offset(facing);
+			TileEntity targetTile = world.getTileEntity(target);
+			boolean opposedChute = targetTile instanceof ChuteTileEntity&&((ChuteTileEntity)targetTile).diagonal;
+			// We're not going to redirect into another diagonal chute, to avoid infinite bouncing
+			if(opposedChute)
+				return;
+
+			long time = System.currentTimeMillis();
+			String hash = Integer.toHexString(getPos().hashCode());
```

Comment:
This seems strange. The only reason for using a hash (that I can think of) would be reduced memory usage in the NBT, but then turning it into a string will probably use more memory, not less. I'd suggest storing `BlockPos#toLong`, since that is unique for vanilla (not for CubicChunks, but still quite good).

Reply:
I wanna say toLong wasn't around in 1.12, but I may be wrong. I'll change this though, for uniformity.